### PR TITLE
Merge freq array function and new frequency dictionary builder

### PR DESCRIPTION
**Pull request by mike-w-wilson**
This adds the ability to merge multiple frequency arrays that are on the same table. The merge supports addition or subtraction of arrays. For subtraction, all freq arrays are subtracted from the first freq array in the list. This also adds a new function to build frequency array dictionaries from the frequency metadata.

### CODE REVIEW

File: gnomad/utils/annotations.py [hunk 1]
```diff
@@ -1106,7 +1109,7 @@ def region_flag_expr(
     :return: `region_flag` struct row annotation
     """
     prob_flags_expr = (
-        {"non_par": (t.locus.in_x_nonpar() | t.locus.in_y_nonpar())} if non_par else {}
+        {"non_par": t.locus.in_x_nonpar() | t.locus.in_y_nonpar()} if non_par else {}
```
Comment (mike-w-wilson) [inline gnomad/utils/annotations.py:1112]:
```suggestion
        {"non_par": (t.locus.in_x_nonpar() | t.locus.in_y_nonpar())} if non_par else {}
```

File: gnomad/utils/annotations.py [hunk 2]
```diff
@@ -125,7 +126,7 @@ def project_max_expr(
                     > 0
                 ),
                 # order the callstats computed by AF in decreasing order
-                lambda x: -x[1].AF[ai]
+                lambda x: -x[1].AF[ai],
```
Comment (mike-w-wilson) [inline gnomad/utils/annotations.py:129]:
```suggestion
                lambda x: -x[1].AF[ai]
```

File: gnomad/utils/annotations.py [hunk 3]
```diff
@@ -125,7 +125,7 @@ def project_max_expr(
...
```
Comment (mike-w-wilson) [inline gnomad/utils/annotations.py:128]:
```suggestion
                lambda x: -x[1].AF[ai]
```

File: gnomad/utils/annotations.py [hunk 4]
```diff
@@ -1106,7 +1108,7 @@ def region_flag_expr(
     :return: `region_flag` struct row annotation
     """
     prob_flags_expr = (
-        {"non_par": (t.locus.in_x_nonpar() | t.locus.in_y_nonpar())} if non_par else {}
+        {"non_par": t.locus.in_x_nonpar() | t.locus.in_y_nonpar()} if non_par else {}
```
Comment (mike-w-wilson) [inline gnomad/utils/annotations.py:1111]:
```suggestion
        {"non_par": (t.locus.in_x_nonpar() | t.locus.in_y_nonpar())} if non_par else {}
```

File: gnomad/utils/annotations.py (hunk 3 above)
Comment (mike-w-wilson) [inline gnomad/utils/annotations.py:128]:
```suggestion
                lambda x: -x[1].AF[ai]
```

File: gnomad/utils/annotations.py [hunk 5]
```diff
@@ -1188,3 +1190,132 @@ def hemi_expr(
...
+        ),
+        fmeta[0],
+        fmeta[1:],
+    )
+
+    # Create a list of tuples from the dictionary, sorted by the list of indices for
+    # each aggregation group.
+    fmeta = hl.sorted(fmeta.items(), key=lambda f: f[1])
+
+    # Create a list of the aggregation groups, maintaining the sorted order
```
Comment (jkgoodrich) [inline gnomad/utils/annotations.py:1259]:
```suggestion
    # Create a list of the aggregation groups, maintaining the sorted order.
```

File: gnomad/utils/annotations.py [hunk 6]
```diff
@@ -1188,3 +1190,132 @@ def hemi_expr(
...
+    new_freq_meta = fmeta.map(lambda x: x[0])
+
+    # Create array for each aggregation group of arrays containing the group's freq
+    # values from each freq array
```
Comment (jkgoodrich) [inline gnomad/utils/annotations.py:1263]:
```suggestion
    # values from each freq array.
```

File: gnomad/utils/annotations.py [hunk 7]
```diff
@@ -1188,3 +1190,132 @@ def hemi_expr(
...
+        :param field_2_expr: Second field to sum or diff.
+        :return: Merged field value.
+        """
+        return hl.if_else(
+            operation == "sum",
+            hl.or_else(field_1_expr, 0) + hl.or_else(field_2_expr, 0),
+            hl.or_else(field_1_expr, 0) - hl.or_else(field_2_expr, 0),
+        )
+
+    # Iterate through the groups and their freq lists to merge callstats
```
Comment (jkgoodrich) [inline gnomad/utils/annotations.py:1282]:
```suggestion
    # Iterate through the groups and their freq lists to merge callstats,
```

File: gnomad/utils/annotations.py [hunk 8]
```diff
@@ -1188,3 +1190,132 @@ def hemi_expr(
...
+    new_freq = freq_meta_idx.map(
```
Comment (jkgoodrich) [inline gnomad/utils/annotations.py:1283]: Maybe this (and below suggestions to reduce redundancy):

```suggestion
    callstat_ann = ["AC", "AN", "homozygote_count"]
    new_freq = freq_meta_idx.map(
```

File: gnomad/utils/annotations.py [hunk 9]
```diff
@@ -1188,3 +1190,132 @@ def hemi_expr(
...
+        lambda x: hl.bind(
+            lambda y: y.annotate(AF=hl.if_else(y.AN > 0, y.AC / y.AN, 0)),
+            hl.fold(
+                lambda i, j: hl.struct(
+                    AC=_sum_or_diff_fields(i.AC, j.AC),
+                    AN=_sum_or_diff_fields(i.AN, j.AN),
+                    homozygote_count=_sum_or_diff_fields(
+                        i.homozygote_count, j.homozygote_count
+                    ),
```
Comment (jkgoodrich) [inline gnomad/utils/annotations.py:1292]:
```suggestion
                    **{
                        ann: _sum_or_diff_fields(i[ann], j[ann])
                        for ann in callstat_ann
                    }
```

File: gnomad/utils/annotations.py [hunk 10]
```diff
@@ -1188,3 +1190,132 @@ def hemi_expr(
...
+                AC=hl.case()
+                .when(set_negatives_to_zero, hl.max(x.AC, 0))
+                .or_error(negative_value_error_msg),
+                AN=hl.case()
+                .when(set_negatives_to_zero, hl.max(x.AN, 0))
+                .or_error(negative_value_error_msg),
+                homozygote_count=hl.case()
+                .when(set_negatives_to_zero, hl.max(x.homozygote_count, 0))
+                .or_error(negative_value_error_msg),
+            )
```
Comment (jkgoodrich) [inline gnomad/utils/annotations.py:1318]: **{
                    ann: hl.case()
                           .when(set_negatives_to_zero, hl.max(x[ann], 0))
                           .or_error(negative_value_error_msg)
                    for ann in callstat_ann
                }

File: gnomad/utils/annotations.py [hunk 11]
```diff
@@ -1188,3 +1190,132 @@ def hemi_expr(
...
+    )
+    # Check and see if any annotation within the merged array is negative. If so,
+    # raise an error if set_negatives_to_zero is False or set the value to 0 if
+    # set_negatives_to_zero is True.
+    if operation == "diff":
+        negative_value_error_msg = (
+            "Negative values found in merged frequency array. Review data or set"
+            " `set_negatives_to_zero` to True to set negative values to 0."
+        )
+        new_freq = new_freq.map(
```
Comment (jkgoodrich) [inline gnomad/utils/annotations.py:1307]: Need to adjust the frequency as well

File: gnomad/utils/release.py [hunk 12]
```diff
@@ -92,3 +94,34 @@ def _get_index(label_groups):
...
+
+
+def make_freq_index_dict_from_meta(
+    freq_meta: List[Dict[str, str]],
+    label_delimiter: str = "_",
+    sort_order: Optional[List[str]] = SORT_ORDER,
+) -> Dict[str, int]:
+    """
+    Create a dictionary for accessing frequency array.
+
```
Comment (jkgoodrich) [inline gnomad/utils/release.py:106]: Add what the key and values are

File: gnomad/utils/release.py [hunk 13]
```diff
@@ -92,3 +94,34 @@ def _get_index(label_groups):
...
+    :param freq_meta: List of dictionaries containing frequency metadata.
+    :param label_delimiter: Delimiter to use when joining frequency metadata labels.
+    :param sort_order: List of frequency metadata labels to use when sorting the dictionary.
+    :return: Dictionary of frequency metadata.
+    """
+    index_dict = {}
```
Comment (jkgoodrich) [inline gnomad/utils/release.py:112]: Maybe add a warning if there are labels in freq_meta that are not in sort_order and if there are labels in sort_order that are not in freq_meta

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
A couple suggestions

File: gnomad/utils/annotations.py (hunk 4 above)
Comment (mike-w-wilson) [inline gnomad/utils/annotations.py:1111]: I'm not sure how to stop black from doing this...

File: gnomad/utils/annotations.py (hunk 3 above)
Comment (mike-w-wilson) [inline gnomad/utils/annotations.py:128]: I think this is actually fine though useless.

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
A couple more things

File: gnomad/utils/annotations.py [hunk 14]
```diff
@@ -1188,3 +1190,131 @@ def hemi_expr(
...
+        will be the union of groupings found within the arrays' metadata and all arrays
+        with be summed by grouping. If `operation` is set to "diff", the merged array
+        will contain groups only found in the first array of `fmeta`. Any array containing
+        any of these groups will have thier values subtracted from the values of the first array.
+
+    :param farrays: List of frequency arrays to merge. First entry in the list is the primary array to which other arrays will be added or subtracted. All arrays must be on the same Table.
+    :param fmeta: List of frequency metadata for arrays being merged.
+    :param operation: Merge operation to perform. Options are "sum" and "diff". If "diff" is passed, the first freq array in the list will have the other arrays subtracted from it.
+    :param set_negatives_to_zero: If True, set negative array values to 0 for AC, AN, AF, and homozygote_count. If False, raise a ValueError. Default is True.
+    :return: Tuple of merged frequency array and its freq index dictionary.
```
[the remark this answers is not in the document]
Comment (mike-w-wilson) [reply gnomad/utils/annotations.py:1225]: 😶‍🌫️

File: gnomad/utils/annotations.py [hunk 15]
```diff
@@ -1106,7 +1108,8 @@ def region_flag_expr(
     :return: `region_flag` struct row annotation
     """
     prob_flags_expr = (
-        {"non_par": (t.locus.in_x_nonpar() | t.locus.in_y_nonpar())} if non_par else {}
+        {"non_par": (t.locus.in_x_nonpar() | t.locus.in_y_nonpar())
+         } if non_par else {}  # fmt: skip
```
Comment (mike-w-wilson) [inline gnomad/utils/annotations.py:1112]: Ha the extra note made the line long enough that black wants parentheses there now...

Comment (jkgoodrich) [reply gnomad/utils/annotations.py:1112]: oh boy. I'm surprised it was even removing the parentheses, mine doesn't do that

**Review by jkgoodrich — APPROVED** [review]
LGTM!

**Pull request merged**